File: 5 kyu/Greed is Good/solution.rb
SCORE_MAP = {
  1 => [0, 100, 200, 1000, 1100, 1200, 2000],
  2 => [0, 0, 0, 200, 200, 200, 400],
  3 => [0, 0, 0, 300, 300, 300, 600],
  4 => [0, 0, 0, 400, 400, 400, 800],
  5 => [0, 50, 100, 500, 550, 600, 1000],
  6 => [0, 0, 0, 600, 600, 600, 1200]
}

def score( dice )
  (1..6).inject(0) do |score, die|
    score += SCORE_MAP[die][dice.count(die)]
  end
end
_____________________________________________
def score( dice )
  [ dice.count(1) / 3 * 1000,
    dice.count(6) / 3 * 600,
    dice.count(5) / 3 * 500,
    dice.count(4) / 3 * 400,
    dice.count(3) / 3 * 300,
    dice.count(2) / 3 * 200,
    dice.count(1) % 3 * 100,
    dice.count(5) % 3 * 50 ].reduce(:+)
end
_____________________________________________
def score( dice )
  m = {1 => 100, 5 => 50}
  (1..6).reduce(0) do|res, i|
    count = dice.count(i)    
    res + count/3 * i * (i==1 ? 1000 : 100) + count%3*(m[i].to_i)
  end
end
_____________________________________________
def score(dice)
  dice.sort.join.gsub(/(\d)\1\1|(1|5)/).inject(0) do |sum, num|
    sum + ($1.to_i * 100 + $2.to_i * 10 ) * (num[0] == '1' ? 10 : 1)
  end
end
_____________________________________________
GREED_SCORES = {
  1 => [0, 100, 200, 1000, 1100, 1200],
  2 => [0, 0, 0, 200, 200, 200],
  3 => [0, 0, 0, 300, 300, 300],
  4 => [0, 0, 0, 400, 400, 400],
  5 => [0, 50, 100, 500, 550, 600],
  6 => [0, 0, 0, 600, 600, 600]
}

def score(dice)
  GREED_SCORES.keys.inject(0) do |score, key|
    score + GREED_SCORES[key][dice.count(key)]
  end
end

File: 5 kyu/Greed is Good/solution.js
function score( dice ) {
  var dc = [0,0,0,0,0,0];
  var tdr = [1000,200,300,400,500,600];
  var sdr = [100,0,0,0,50,0];
  dice.forEach(function(x){ dc[x-1]++; });
  return dc.reduce(function(s,x,i){ 
    return s + (x >= 3? tdr[i] : 0) + sdr[i]*(x % 3);
  },0);
}
_____________________________________________
function score( dice ) {
  var six=0, five=0, four=0, three=0, too=0, one=0;
  var i = 0;
  while (i < 5) {
    if (dice[i] == 6) { six++; }
    if (dice[i] == 5) { five++; }
    if (dice[i] == 4) { four++; }
    if (dice[i] == 3) { three++; }
    if (dice[i] == 2) { too++; }
    if (dice[i] == 1) { one++; }
    i++;
  }
  var r = 0;
  if (one > 2) { r += 1000; one -= 3;}
  if (six > 2) { r += 600; }
  if (five > 2) { r += 500; five -= 3;}
  if (four > 2) { r += 400; }
  if (three > 2) { r += 300; }
  if (too > 2) { r += 200; }
  r += one * 100;
  r += five * 50;
  return r;
}
_____________________________________________
function score( dice ) {
  if (dice.length !== 5) return 0;
  
  let diceStr = dice.sort().join('');
  let score = 0;
  const rules = [
    { reg: /111/, score: 1000 },
    { reg: /666/, score: 600 },
    { reg: /555/, score: 500 },
    { reg: /444/, score: 400 },
    { reg: /333/, score: 300 },
    { reg: /222/, score: 200 },
    { reg: /1/,   score: 100 },
    { reg: /5/,   score: 50 },
  ];
  
  rules.forEach(rule => {
    while (rule.reg.test(diceStr)) {
      diceStr = diceStr.replace(rule.reg, '');
      score += rule.score;
    }
  });
  
  return score;
}
_____________________________________________
function score( dice ) {
  var score = [0, 0, 0, 0, 0, 0];

  dice.forEach(function(die) {
    ++score[die - 1];
  });

  return score.reduce(function(total, n, i) {
    switch (i + 1) {
      case 1:
        return total + (Math.floor(n / 3) * 1000) + ((n % 3) * 100);

      case 5:
        return total + (Math.floor(n / 3) * 500) + ((n % 3) * 50);

      default:
        return total + (Math.floor(n / 3) * (i + 1) * 100);
    }
  }, 0);
}
_____________________________________________
function score(dice) {
  var sum = 0;
  dice = dice.sort();
  for(var i = 0; i < dice.length; i++){
    if (dice[i] == dice[i + 1] && dice[i + 1] == dice[i + 2]) {
      switch(dice[i]) {
        case 1:
          sum += 1000; break;
        case 6:
          sum += 600; break;
        case 5:
          sum += 500; break;
        case 4:
          sum += 400; break;
        case 3:
          sum += 300; break;
        case 2:
          sum += 200;
      }
      i = i + 2;
      continue;
      }
    else if (dice[i] == 1) {
      sum += 100;
    } else if (dice[i] == 5) {
      sum += 50;
    }
  }
  return sum;
}
